flytekit/core/python_auto_container.py

```diff
@@ -273,6 +274,37 @@ def get_all_tasks(self) -> List[PythonAutoContainerTask]:  # type: ignore
 default_task_resolver = DefaultTaskResolver()
 
 
+def update_image_spec_copy_handling(image_spec: ImageSpec, settings: SerializationSettings):
+    """
+    This helper function is where the relationship between fast register and ImageSpec is codified.
+    If fast register is not enabled, then source root is used and then files are copied.
+    See the copy option in ImageSpec for more information.
+
+    Currently the relationship is incidental. Because serialization settings are not passed into the image spec
+    build command (and it probably shouldn't be), the builder has no concept of which files to copy, when, and
+    from where. (or to where but that is hard-coded)
+    """
+    # Handle when the copy method is explicitly set by the user.
+    if image_spec.copy is not None:
+        if image_spec.copy != CopyFileDetection.NO_COPY:
+            # if we need to copy any files, make sure source root is set. This preserves the behavior pre-copy arg,
+            # and allows the user to not have to specify source root.
+            if image_spec.source_root is None and settings.source_root is not None:
+                image_spec.source_root = settings.source_root
+
+    # Handle the default behavior of setting the behavior based on the inverse of fast register usage
+    # The default behavior additionally requires that serializa
+    elif settings.fast_serialization_settings is None or not settings.fast_serialization_settings.enabled:
+        # Set the source root for the image spec if it's non-fast registration
+        # Unfortunately whether the source_root/copy instructions should be set is implicitly dependent also on the
+        # existence of the source root in settings.
+        if settings.source_root is not None or image_spec.source_root is not None:
+            if image_spec.source_root is None:
+                image_spec.source_root = settings.source_root
```

Comment:
```suggestion
              image_spec.source_root = settings.source_root
```

Reply:
why?  what if it's set manually by the user in the constructor of ImageSpec?

Reply:
and if users are not supposed to set that then we should remove it from the constructor.  If users are allowed to set it, we should add a comment that says this means it'll rebuild every time those files change.

Reply:
oops, sorry. my bad. I thought line 301 was
```
if settings.source_root is not None and image_spec.source_root is not None:
```